crates/iceberg/src/spec/manifest_list.rs

```diff
@@ -49,77 +57,128 @@ impl ManifestList {
-        ];
-        Schema::builder().with_fields(fields).build().unwrap()
+/// A manifest list writer.
+pub struct ManifestListWriter {
+    format_version: FormatVersion,
+    output_file: OutputFile,
+    avro_writer: Writer<'static, Vec<u8>>,
+}
+
+impl ManifestListWriter {
+    /// Construct a v1 [`ManifestListWriter`] that writes to a provided [`OutputFile`].
+    pub fn v1(output_file: OutputFile, snapshot_id: i64, parent_snapshot_id: i64) -> Self {
+        let mut metadata = HashMap::new();
+        metadata.insert("snapshot-id".to_string(), snapshot_id.to_string());
+        metadata.insert(
+            "parent-snapshot-id".to_string(),
+            parent_snapshot_id.to_string(),
+        );
+        metadata.insert("format-version".to_string(), "1".to_string());
+        Self::new(FormatVersion::V1, output_file, metadata)
     }
 
-    /// Get the v1 schema of the manifest list entry.
-    pub(crate) fn v1_schema() -> Schema {
-        let fields = vec![
-            _const_fields::MANIFEST_PATH.clone(),
-            _const_fields::MANIFEST_LENGTH.clone(),
-            _const_fields::PARTITION_SPEC_ID.clone(),
-            _const_fields::ADDED_SNAPSHOT_ID.clone(),
-            _const_fields::ADDED_FILES_COUNT_V1.clone().to_owned(),
-            _const_fields::EXISTING_FILES_COUNT_V1.clone(),
-            _const_fields::DELETED_FILES_COUNT_V1.clone(),
-            _const_fields::ADDED_ROWS_COUNT_V1.clone(),
-            _const_fields::EXISTING_ROWS_COUNT_V1.clone(),
-            _const_fields::DELETED_ROWS_COUNT_V1.clone(),
-            _const_fields::PARTITIONS.clone(),
-            _const_fields::KEY_METADATA.clone(),
-        ];
-        Schema::builder().with_fields(fields).build().unwrap()
+    /// Construct a v2 [`ManifestListWriter`] that writes to a provided [`OutputFile`].
+    pub fn v2(
+        output_file: OutputFile,
+        snapshot_id: i64,
+        parent_snapshot_id: i64,
+        sequence_number: i64,
+    ) -> Self {
+        let mut metadata = HashMap::new();
+        metadata.insert("snapshot-id".to_string(), snapshot_id.to_string());
+        metadata.insert(
+            "parent-snapshot-id".to_string(),
+            parent_snapshot_id.to_string(),
+        );
+        metadata.insert("sequence-number".to_string(), sequence_number.to_string());
+        metadata.insert("format-version".to_string(), "2".to_string());
+        Self::new(FormatVersion::V2, output_file, metadata)
+    }
+
+    fn new(
+        format_version: FormatVersion,
+        output_file: OutputFile,
+        metadata: HashMap<String, String>,
+    ) -> Self {
+        let avro_schema = match format_version {
+            FormatVersion::V1 => &MANIFEST_LIST_AVRO_SCHEMA_V1,
+            FormatVersion::V2 => &MANIFEST_LIST_AVRO_SCHEMA_V2,
+        };
+        let mut avro_writer = Writer::new(avro_schema, Vec::new());
+        for (key, value) in metadata {
+            avro_writer.add_user_metadata(key, value).unwrap();
```

Comment:
How about using `expect()` to clarify why we anticipate this operation will always succeed?